pkg/kubectl/cmd/exec.go

```diff
@@ -44,96 +45,144 @@ $ kubectl exec 123456-7890 -c ruby-container -i -t -- bash -il`
+			cmdutil.CheckErr(options.Run())
 		},
 	}
-	cmd.Flags().StringVarP(&params.podName, "pod", "p", "", "Pod name")
+	cmd.Flags().StringVarP(&options.PodName, "pod", "p", "", "Pod name")
 	// TODO support UID
-	cmd.Flags().StringVarP(&params.containerName, "container", "c", "", "Container name")
-	cmd.Flags().BoolVarP(&params.stdin, "stdin", "i", false, "Pass stdin to the container")
-	cmd.Flags().BoolVarP(&params.tty, "tty", "t", false, "Stdin is a TTY")
+	cmd.Flags().StringVarP(&options.ContainerName, "container", "c", "", "Container name")
+	cmd.Flags().BoolVarP(&options.Stdin, "stdin", "i", false, "Pass stdin to the container")
+	cmd.Flags().BoolVarP(&options.TTY, "tty", "t", false, "Stdin is a TTY")
 	return cmd
 }
 
-type remoteExecutor interface {
+// RemoteExecutor defines the interface accepted by the Exec command - provided for test stubbing
+type RemoteExecutor interface {
 	Execute(req *client.Request, config *client.Config, command []string, stdin io.Reader, stdout, stderr io.Writer, tty bool) error
 }
 
-type defaultRemoteExecutor struct{}
+// DefaultRemoteExecutor is the standard implementation of remote command execution
+type DefaultRemoteExecutor struct{}
 
-func (*defaultRemoteExecutor) Execute(req *client.Request, config *client.Config, command []string, stdin io.Reader, stdout, stderr io.Writer, tty bool) error {
+func (*DefaultRemoteExecutor) Execute(req *client.Request, config *client.Config, command []string, stdin io.Reader, stdout, stderr io.Writer, tty bool) error {
 	executor := remotecommand.New(req, config, command, stdin, stdout, stderr, tty)
 	return executor.Execute()
 }
 
-type execParams struct {
-	podName       string
-	containerName string
-	stdin         bool
-	tty           bool
+// ExecOptions declare the arguments accepted by the Exec command
+type ExecOptions struct {
+	Namespace     string
+	PodName       string
+	ContainerName string
+	Stdin         bool
+	TTY           bool
+	Command       []string
+
+	In  io.Reader
+	Out io.Writer
+	Err io.Writer
+
+	Executor RemoteExecutor
+	Client   *client.Client
+	Config   *client.Config
 }
 
-func extractPodAndContainer(cmd *cobra.Command, argsIn []string, p *execParams) (podName string, containerName string, args []string, err error) {
-	if len(p.podName) == 0 && len(argsIn) == 0 {
-		return "", "", nil, cmdutil.UsageError(cmd, "POD is required for exec")
+// Complete verifies command line arguments and loads data from the command environment
+func (p *ExecOptions) Complete(f *cmdutil.Factory, cmd *cobra.Command, argsIn []string) error {
+	if len(p.PodName) == 0 && len(argsIn) == 0 {
+		return cmdutil.UsageError(cmd, "POD is required for exec")
 	}
-	if len(p.podName) != 0 {
+	if len(p.PodName) != 0 {
 		printDeprecationWarning("exec POD", "-p POD")
-		podName = p.podName
 		if len(argsIn) < 1 {
-			return "", "", nil, cmdutil.UsageError(cmd, "COMMAND is required for exec")
+			return cmdutil.UsageError(cmd, "COMMAND is required for exec")
```

Comment:
Double validates because this is specific to command behavior (Complete)
not usage behavior (Validate)

On Thu, Jul 23, 2015 at 3:43 PM, David Eads <notifications@github.com>
wrote:

> In pkg/kubectl/cmd/exec.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11729#discussion_r35362768>
> :
>
> >  		if len(argsIn) < 1 {
> > -			return "", "", nil, cmdutil.UsageError(cmd, "COMMAND is required for exec")
> > +			return cmdutil.UsageError(cmd, "COMMAND is required for exec")
>
> seems like validation best left to Validate
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11729/files#r35362768>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift